### PR TITLE
kip-114: Add RANDAO header fields

### PR DESCRIPTION
**Pull request by blukat29**
## Proposed changes

- Add header fields `randomReveal` and `mixHash`
- Add predicates `ChainConfig.IsRandaoForkEnabled` and `Rules.IsRandao`
- Following APIs return the new fields:
  - `eth_getHeader*` `eth_getBlock*` returns nonzero `mixHash` after Randao fork (was 0x00..00 before fork)
  - `klay_getHeader*` `klay_getBlock*` returns `randomReveal` and `mixHash` after Randao fork 

## Types of changes

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

Future works
- Istanbul consensus engine processes Randao fields
- EVM Random opcode
- Proposer selection based on Randao fields

### CODE REVIEW

File: blockchain/types/block.go [hunk 1]
```diff
@@ -63,21 +63,32 @@ type Header struct {
 	Governance []byte `json:"governanceData"            gencodec:"required"`
 	Vote       []byte `json:"voteData,omitempty"`
 
-	BaseFee *big.Int `json:"baseFeePerGas,omitempty"    rlp:"optional"`
+	// Added with Magma hardfork for KIP-71 dynamic fee.
+	BaseFee *big.Int `json:"baseFeePerGas,omitempty" rlp:"optional"`
+
+	// Added with Randao hardfork for KIP-114 RANDAO.
+	RandomReveal []byte `json:"randomReveal,omitempty" rlp:"optional"` // 96 byte BLS signature
```
Comment (hyunsooda) [inline blockchain/types/block.go:70]: Why not use array `[96]byte` rather slice. Or, a new type may be introduced like `common.hash`

File: blockchain/types/block.go [hunk 2]
```diff
@@ -63,21 +63,32 @@ type Header struct {
...
+	MixHash      []byte `json:"mixHash,omitempty" rlp:"optional"`      // 32 byte RANDAO mix
```
Comment (hyunsooda) [inline blockchain/types/block.go:71]: `common.hash`?

Comment (blukat29) [reply blockchain/types/block.go:71]: I used `[]byte` to distinguish between 'field not defined' and 'zero hash'.
Same for the RandomReveal.

File: api/api_public_blockchain.go [hunk 3]
```diff
@@ -574,13 +576,17 @@ func RpcOutputBlock(b *types.Block, td *big.Int, inclTx bool, fullTx bool, isEna
...
-	if isEnabledEthTxTypeFork {
+	if rules.IsEthTxType {
 		if head.BaseFee == nil {
 			fields["baseFeePerGas"] = (*hexutil.Big)(new(big.Int).SetUint64(params.ZeroBaseFee))
 		} else {
 			fields["baseFeePerGas"] = (*hexutil.Big)(head.BaseFee)
 		}
 	}
+	if rules.IsRandao {
+		fields["randomReveal"] = hexutil.Bytes(head.RandomReveal)
```
Comment (hyunsooda) [inline api/api_public_blockchain.go:587]: field name typo

File: blockchain/types/block.go (hunk 2 above)
Comment (hyunsooda) [reply blockchain/types/block.go:71]: `RandomReveal != nil` is the same format with `RandomReveal != (common.RandomReveal{})`. 
I know that you prefer the first case. However, we have a potential error case against mismatched type, and it brings length check in almost usage. Could you reconsider it again?

Comment (blukat29) [reply blockchain/types/block.go:71]: It's not a syntax issue. `[96]byte` and `[]byte` behaves differently in RLP encoding.

Recall that RLP encodes an array. The `rlp:"optional"` field may not present in the encoding, making the array short.
If we decode a short array into Header struct, `[96]byte` will be filled with 96 zero bytes, whereas `[]byte` will be nil. Refer to the test code below.

I need to clearly distinguish 'field not exists' and 'zero value' in consensus code as [we did with BaseFee field](https://github.com/klaytn/klaytn/blob/dev/consensus/istanbul/backend/engine.go#L206).

<details>
<summary>Details</summary>

```go
func TestHello(t *testing.T) {
	root := common.HexToHash("0xce26cd14bfe586c11f9ebae0186a3e4744dfb6233e603d477cdebc69d9dc7953")
	_ = root

	type Reveal [96]byte

	type A struct {
		Root   common.Hash
		Reveal Reveal      `rlp:"optional"`
		Mix    common.Hash `rlp:"optional"`
	}

	type B struct {
		Root   common.Hash
		Reveal []byte `rlp:"optional"`
		Mix    []byte `rlp:"optional"`
	}

	enc := func(x interface{}) {
		r, _ := rlp.EncodeToBytes(x)
		fmt.Printf("%x\n", r)
	}
	decA := func(rs string) {
		var x A
		rlp.DecodeBytes(hexutil.MustDecode(rs), &x)
		fmt.Printf("A{%x %x %x}\n", x.Root, x.Reveal, x.Mix)
	}
	decB := func(rs string) {
		var x B
		rlp.DecodeBytes(hexutil.MustDecode(rs), &x)
		fmt.Printf("B{%x %x %x}\n", x.Root, x.Reveal, x.Mix)
	}
	_, _, _ = enc, decA, decB

	decA("0xe1a0ce26cd14bfe586c11f9ebae0186a3e4744dfb6233e603d477cdebc69d9dc7953")
	decA("0xf8a4a0ce26cd14bfe586c11f9ebae0186a3e4744dfb6233e603d477cdebc69d9dc7953b860000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000000")

	decB("0xe1a0ce26cd14bfe586c11f9ebae0186a3e4744dfb6233e603d477cdebc69d9dc7953")
	decB("0xf8a4a0ce26cd14bfe586c11f9ebae0186a3e4744dfb6233e603d477cdebc69d9dc7953b860000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000000")

	//enc(A{root, Reveal{}, common.Hash{}})
	//enc(B{root, nil, nil})
	//enc(B{root, make([]byte, 96), make([]byte, 32)})
}
```

</details>

File: node/cn/filters/api.go [hunk 4]
```diff
@@ -243,13 +244,17 @@ func RPCMarshalHeader(head *types.Header, isEnabledEthTxTypeFork bool) map[strin
...
-	if isEnabledEthTxTypeFork {
+	if rules.IsEthTxType {
 		if head.BaseFee == nil {
 			result["baseFeePerGas"] = (*hexutil.Big)(new(big.Int).SetUint64(params.ZeroBaseFee))
 		} else {
 			result["baseFeePerGas"] = (*hexutil.Big)(head.BaseFee)
 		}
 	}
+	if rules.IsRandao {
+		result["reandomReveal"] = hexutil.Bytes(head.RandomReveal)
```
Comment (hyunsooda) [inline node/cn/filters/api.go:255]: I meant this. Sorry for the wrong pointer.
```suggestion
		result["randomReveal"] = hexutil.Bytes(head.RandomReveal)
```

File: blockchain/types/block.go (hunk 2 above)
Comment (hyunsooda) [reply blockchain/types/block.go:71]: It makes sense. Got it, thank you.

File: node/cn/filters/api.go (hunk 4 above)
Comment (blukat29) [reply node/cn/filters/api.go]: Fixed. Thank you!

**Review by hyunsooda — APPROVED** [review]

**Review by yoomee1313 — DISMISSED** [review]

**Comment by blukat29** [comment]
@aidan-kwon @ian0371 PTAL

**Review by hyeonLewis — DISMISSED** [review]

File: blockchain/types/gen_header_json.go [hunk 5]
```diff
@@ -16,22 +16,24 @@ var _ = (*headerMarshaling)(nil)
...
+		BlockScore   *hexutil.Big   `json:"blockScore"       gencodec:"required"`
+		Number       *hexutil.Big   `json:"number"           gencodec:"required"`
+		GasUsed      hexutil.Uint64 `json:"gasUsed"          gencodec:"required"`
+		Time         *hexutil.Big   `json:"timestamp"        gencodec:"required"`
+		TimeFoS      hexutil.Uint   `json:"timestampFoS"              gencodec:"required"`
+		Extra        hexutil.Bytes  `json:"extraData"                 gencodec:"required"`
+		Governance   hexutil.Bytes  `json:"governanceData"            gencodec:"required"`
+		Vote         hexutil.Bytes  `json:"voteData,omitempty"`
+		BaseFee      *hexutil.Big   `json:"baseFeePerGas,omitempty" rlp:"optional"`
+		RandomReveal []byte         `json:"randomReveal,omitempty" rlp:"optional"`
```
Comment (ian0371) [inline blockchain/types/gen_header_json.go:34]: Why not follow the convention `hexutil.Bytes`?

Comment (blukat29) [reply blockchain/types/gen_header_json.go]: Regenerated gen_header_json.

**Review by hyeonLewis — DISMISSED** [review]

**Review by ian0371 — APPROVED** [review]

**Comment by blukat29** [comment]
Now linter should pass. Sorry for the hassle.

**Review by hyeonLewis — APPROVED** [review]

**Review by hyunsooda — APPROVED** [review]

**Review by aidan-kwon — APPROVED** [review]

**Pull request merged**